addon/PageScraper.js

```diff
@@ -30,33 +31,64 @@ function PageScraper(previewProvider, tabTracker, options = {}) {
 
 PageScraper.prototype = {
   /**
-   * Check if the link is already in the metadata database. If not, parse the
-   * HTML and insert it in the metadata database
+   * Parse the HTML and attempt to insert it in the metadata database
    */
   _asyncParseAndSave: Task.async(function*(rawHTML, url, event) {
-    let startTime = Date.now();
-    this._tabTracker.handlePerformanceEvent(event, "metadataReceivedRawHTML", startTime);
-
-    let link = yield this._previewProvider.asyncLinkExist(url);
-    if (!link) {
+    this._tabTracker.handlePerformanceEvent(event, "metadataReceivedRawHTML", Date.now());
+    let linkExists = yield this._previewProvider.asyncLinkExist(url);
+    if (!linkExists) {
       const metadata = this._metadataParser.parseHTMLText(rawHTML, url);
-      if (this._shouldSaveMetadata(metadata)) {
-        try {
-          metadata.images = yield this._computeImageSize(metadata);
-        } catch (e) {
-          Cu.reportError(`PageScraper failed to compute image size for ${url}`);
-        }
-        this._insertMetadata(metadata);
-        this._tabTracker.handlePerformanceEvent(event, "metadataParseSuccess", Date.now() - startTime);
-      } else {
-        this._tabTracker.handlePerformanceEvent(event, "metadataInvalidReceived", Date.now() - startTime);
+      this._asyncSaveMetadata(metadata, event);
+    } else {
+      this._tabTracker.handlePerformanceEvent(event, "metadataExists", Date.now());
+    }
+  }),
+
+  /**
+   * Save the metadata in the MetadataStore DB
+   */
+  _asyncSaveMetadata: Task.async(function*(metadata, event) {
+    const startTime = Date.now();
+    if (this._shouldSaveMetadata(metadata)) {
```

Comment:
Fail fast out of this if statement instead of embedding the entire function body inside of it